### PR TITLE
Windows: Enable, test, and then disable SoundSourceFFmpeg as fallback

### PR DESCRIPTION
**Pull request by uklotzde**
Caveat: On Windows no version numbers are reported!

### CODE REVIEW

File: CMakeLists.txt
Comment (daschuer) [inline CMakeLists.txt:2304]: Is this a bug in our windows build environment?

Comment (Be-ing) [reply CMakeLists.txt:2304]: perhaps a bug in vcpkg

**Comment by uklotzde** [comment]
Opinions? Just some minor fixes and a nice fallback that should not cause any issues. All available SoundSources are tested, all green.

**Comment by daschuer** [comment]
I think this is a good idea. I just cannot approve this because of absence of a test system.

**Comment by uklotzde** [comment]
> I think this is a good idea. I just cannot approve this because of absence of a test system.

This cannot be tested without tweaking the priorities in a custom build. SoundSourceFFmpeg will probably never be used.

**Comment by uklotzde** [comment]
```
Debug [Main]: SoundSourceProxy - Registering SoundSource providers
Debug [Main]: SoundSourceProxy - Registering platform and fallback SoundSource providers
Debug [Main]: SoundSourceProviderRegistry - Registering provider "Microsoft Media Foundation"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "aac" for provider "Microsoft Media Foundation" with priority 4 (higher)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "m4a" for provider "Microsoft Media Foundation" with priority 4 (higher)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mp4" for provider "Microsoft Media Foundation" with priority 4 (higher)
Debug [Main]: SoundSourceProviderRegistry - Registering provider "FFmpeg"
Info [Main]: SoundSourceFFmpeg - Disabling untested input formats: aa, ac3, acm, act, adf, adp, ads, adx, aea, afc, aix, alp, amr, amrnb, amrwb, anm, apc, ape, apm, apng, aptx, aptx_hd, aqtitle, argo_asf, asf, asf_o, ass, ast, au, av1, avi, avr, avs, avs2, bethsoftvid, bfi, bin, bink, bit, bmv, bfstm, brstm, boa, c93, caf, cavsvideo, cdg, cdxl, cine, codec2, codec2raw, concat, data, daud, dcstr, derf, dfa, dhav, dirac, dnxhd, dsf, dsicin, dss, dts, dtshd, dv, dvbsub, dvbtxt, dxa, ea, ea_cdata, eac3, epaf, ffmetadata, filmstrip, fits, flac, flic, flv, live_flv, 4xm, frm, fsb, fwse, g722, g723_1, g726, g726le, g729, gdv, genh, gif, gsm, gxf, h261, h263, h264, hca, hcom, hevc, hls, hnm, ico, idcin, idf, iff, ifv, ilbc, image2, image2pipe, alias_pix, brender_pix, ingenient, ipmovie, ircam, iss, iv8, ivf, ivr, jacosub, jv, kux, kvag, lmlm4, loas, lrc, lvf, lxf, matroska,webm, mgsts, microdvd, mjpeg, mjpeg_2000, mlp, mlv, mm, mmf, mpc, mpc8, mpeg, mpegts, mpegtsraw, mpegvideo, mpjpeg, mpl2, mpsub, msf, msnwctcp, mtaf, mtv, musx, mv, mvi, mxf, mxg, nc, nistsphere, nsp, nsv, nut, nuv, ogg, oma, paf, alaw, mulaw, vidc, f64be, f64le, f32be, f32le, s32be, s32le, s24be, s24le, s16be, s16le, s8, u32be, u32le, u24be, u24le, u16be, u16le, u8, pjs, pmp, pp_bnk, pva, pvf, qcp, r3d, rawvideo, realtext, redspark, rl2, rm, roq, rpl, rsd, rso, rtp, rtsp, s337m, sami, sap, sbc, sbg, scc, sdp, sdr2, sds, sdx, film_cpk, ser, shn, siff, sln, smk, smjpeg, smush, sol, sox, spdif, srt, psxstr, stl, subviewer1, subviewer, sup, svag, swf, tak, tedcaptions, thp, 3dostr, tiertexseq, tmv, truehd, tta, txd, tty, ty, v210, v210x, vag, vc1, vc1test, vividas, vivo, vmd, vobsub, voc, vpk, vplayer, vqf, w64, wc3movie, webm_dash_manifest, webvtt, wsaud, wsd, wsvqa, wtv, wve, xa, xbin, xmv, xvag, xwma, yop, yuv4mpegpipe, bmp_pipe, dds_pipe, dpx_pipe, exr_pipe, gif_pipe, j2k_pipe, jpeg_pipe, jpegls_pipe, pam_pipe, pbm_pipe, pcx_pipe, pgmyuv_pipe, pgm_pipe, pictor_pipe, png_pipe, ppm_pipe, psd_pipe, qdraw_pipe, sgi_pipe, svg_pipe, sunrast_pipe, tiff_pipe, webp_pipe, xpm_pipe, xwd_pipe
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "aac" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "aif" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "aiff" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "m4v" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mov" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mp4" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "m4a" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "3gp" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "3g2" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mj2" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mp3" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "wav" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "wv" for provider "FFmpeg" with priority 1 (lowest)
Debug [Main]: SoundSourceProviderRegistry - Registering provider "MAD: MPEG Audio Decoder"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mp3" for provider "MAD: MPEG Audio Decoder" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering provider "MODPlug"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "mod" for provider "MODPlug" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "med" for provider "MODPlug" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "okt" for provider "MODPlug" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "s3m" for provider "MODPlug" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "stm" for provider "MODPlug" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "xm" for provider "MODPlug" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "it" for provider "MODPlug" with priority 3 (default)
Info [Main]: SoundSourceSndFile - Disabling OGG decoding for "libsndfile-1.0.31"
Debug [Main]: SoundSourceProviderRegistry - Registering provider "libsndfile"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "aif" for provider "libsndfile" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "aiff" for provider "libsndfile" with priority 3 (default)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "caf" for provider "libsndfile" with priority 2 (lower)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "flac" for provider "libsndfile" with priority 2 (lower)
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "wav" for provider "libsndfile" with priority 3 (default)
Debug [Main]: SoundSourceProxy - Registering reference SoundSource providers
Debug [Main]: SoundSourceProviderRegistry - Registering provider "Xiph.org libFLAC"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "flac" for provider "Xiph.org libFLAC" with priority 4 (higher)
Debug [Main]: SoundSourceProviderRegistry - Registering provider "Xiph.org OggVorbis"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "ogg" for provider "Xiph.org OggVorbis" with priority 4 (higher)
Debug [Main]: SoundSourceProviderRegistry - Registering provider "Xiph.org libopusfile"
Debug [Main]: SoundSourceProviderRegistry - Registering file extension "opus" for provider "Xiph.org libopusfile" with priority 4 (higher)
```

**Comment by Holzhaus** [comment]
So this adds support for obscure file formats like "3gp" (wasn't that used for cell phone videos in the early 2000s?) while leaving the  current decoders untouched? Makes sense to me.

Do you think this will causes legal issues with licenses/patents for US users?

**Comment by uklotzde** [comment]
I am not going to investigate any licensing issues.

Then simply lets set FFMPEG=OFF on Windows and keep the remaining changes. I don't care.

File: CMakeLists.txt
Comment (Holzhaus) [inline CMakeLists.txt:2309]: I think we should stop using `cmake_dependent_option` (except for options that simply don't exist on that OS).

The reason is that the option will forcibly set to off if the condition is not met, regardless of what the user specified.
That means that if I compile Mixxx with `-DFFMPEG=ON - DMAD=OFF` but forgot to install FFMPEG, Mixxx will be built without it but I don't get any warning (and  wonder why MP3 files won't open).

Instead, this should just be a regular option, with a default value "OFF" for now. We can also consider `"${FFMPEG_FOUND}"` as default but I didn't test that.

The version check can go into the FFMPEG block:

```
if(FFMPEG)
  if(NOT FFMPEG_FOUND)
    message(FATAL_ERROR "FFMPEG was not found") 
  endif()
  if(UNIX AND FFMPEG_libavcodec_VERSION VERSION_LESS 58)
    message(FATAL_ERROR "FFMPEG version needs to be >= 58") 
  endif()
```

We can also consider writing a macro for this to reduce the amount of boilerplate code (not in this PR though).

Comment (Holzhaus) [reply CMakeLists.txt:2309]: Instead of a custom if block, we can also repeat the `find_package` call and add `REQUIRED` to the arguments.

**Comment by uklotzde** [comment]
Done here. Please take over if you need more.

File: CMakeLists.txt
Comment (Holzhaus) [inline CMakeLists.txt:2304]: Ubuntu build fails. I suggest:

```suggestion
option(FFMPEG "FFmpeg 4.x support" OFF)
```

**Comment by uklotzde** [comment]
I have added some more unrelated changes mainly to improve consistency.

File: CMakeLists.txt
Comment (Holzhaus) [inline CMakeLists.txt:2625]: I still think this is a bad idea for the reasons I outlined here: https://github.com/mixxxdj/mixxx/pull/3707#discussion_r595508414

I think we should stop using `cmake_dependent_option` and use a regular option instead.

Comment (uklotzde) [reply CMakeLists.txt:2625]: This only establishes consistency. If we want to change it we should change the behavior of **all** feature options **at once**.

Comment (Holzhaus) [reply CMakeLists.txt:2625]: Ok

Comment (uklotzde) [reply CMakeLists.txt:2625]: Inconsistent behavior is the worst of all ;)

File: CMakeLists.txt
Comment (uklotzde) [reply CMakeLists.txt:2304]: No exception for outdated Ubuntu LTS versions. You have to disable FFmpeg explicitly if it is installed. All cmake_dependent_option() features should be consistent now.

File: .github/workflows/build.yml
Comment (daschuer) [inline .github/workflows/build.yml:18]: Please update also the debian/rules file with this setting

Comment (uklotzde) [reply .github/workflows/build.yml:18]: Wait. Then FFmpeg will be disabled for all our Ubuntu packages.

That's why I always argued that the packaging files do not belong in the main repo. We need separate versions for separate distribution releases. It only worked until now because of the hacks in CMakeLists.txt.

Comment (daschuer) [reply .github/workflows/build.yml:18]: Ah yes. Unfortunately this issue blocks this PR. 

Maybe it is pragmatic to restore the special case CMakeList.txt and proceed.

This offers the best out of the box behavior for home builds as well.

Comment (uklotzde) [reply .github/workflows/build.yml:18]: No. It has to be fixed in the `rules`.

I will retarget this PR to `main` or otherwise prefer to close it.

**Comment by uklotzde** [comment]
Rebased on main.

File: .github/workflows/build.yml
Comment (daschuer) [reply .github/workflows/build.yml]: Right, we don't

Comment (Holzhaus) [reply .github/workflows/build.yml:18]: Then we can just disable it on Ubuntu and merge this into 2.3?

**Comment by uklotzde** [comment]
Yet another rebase on 2.3. Over and out.

**Comment by Holzhaus** [comment]
> Rebased on main.

Why? I thought we weren't shipping FFmpeg on Ubuntu anyways?

**Comment by daschuer** [comment]
Ok so we can now disable ffmpeg in the debian/rule file.
A requirement before merging this.

**Comment by uklotzde** [comment]
Corollary: The main repo should only contain the generic Linux packaging stuff. Packaging and deploying to the PPA for Debian/Ubuntu should be moved into a separate repo.

**Comment by Holzhaus** [comment]
I was planning to get rid of the `cmake_dependent_option` stuff in 2.3 but I'm not keep on resolving the merge conflicts with main.

I still don't understand why this PR shouldn't be merge into 2.3.

**Comment by daschuer** [comment]
> Corollary: The main repo should only contain the generic Linux packaging stuff. Packaging and deploying to the PPA for Debian/Ubuntu should be moved into a separate repo.

... to not notice early that a change breaks packaging? And to let package mainter puzzle what we have changed.

I am the opposite opinion. We can make sure that CPack is in a good shape that it is able to generate all distros we maintain. 

There is no got reason why windows and Mac packages are generated directly but not the Linux packages. This way our deployment is always consistent without the ris of package maintainers braking things because nog following our development closely.

**Comment by uklotzde** [comment]
> I though if is enable. Because you explicit disabled it for our CI

...because we install the FFmpeg dependencies by default for the CI builds and I don't check the packaging files all the time. The Debian packaging is not part of the CI build and I am not able to detect any issues.

**Comment by Holzhaus** [comment]
@uklotzde in that case can you reset your branch (using `git reflog` to the state before the rebase)? I think we can merge this into 2.3 now.

**Review by Holzhaus — APPROVED** [review]
LGTM, thanks.

File: CMakeLists.txt
Comment (Holzhaus) [inline CMakeLists.txt:2625]: Oops, I missed this error. I'll replace BULK with WAVPACK in another PR.

Comment (uklotzde) [reply CMakeLists.txt:2625]: Oh, copy & paste error :see_no_evil:

**Pull request merged**
